src/lib.rs

```diff
@@ -1,173 +1,73 @@
+
+    /// Returns the texture coordinate of a vertex.
+    fn get_tex_coord(&self, face: usize, vert: usize) -> Point2<f32>;
+
+    /// Sets a vertex' generated tangent.
+    fn set_tangent(
+        &mut self,
+        tangent: Vector3<f32>,
+        _bi_tangent: Vector3<f32>,
+        _f_mag_s: f32,
+        _f_mag_t: f32,
+        bi_tangent_preserves_orientation: bool,
+        face: usize,
+        vert: usize,
+    ) {
+        let sign = if bi_tangent_preserves_orientation {
+            1.0
+        } else {
+            -1.0
+        };
+        self.set_tangent_encoded(tangent.insert_row(3, sign), face, vert);
     }
-}
 
-/// Returns the position of the referenced face of vertex number
-/// iVert, in the range {0,1,2} for triangles, and {0,1,2,3} for quads.
-extern "C" fn position(
-    pContext: *const ffi::SMikkTSpaceContext,
-    fvPosOut: *mut c_float,
-    iFace: c_int,
-    iVert: c_int,
-) {
-    unsafe {
-        let x = (*pContext).m_pUserData as *const Closures;
-        let slice = ((*x).position)(iFace as usize, iVert as usize);
-        let src = slice.as_ptr() as *const c_float;
-        ptr::copy_nonoverlapping::<c_float>(src, fvPosOut, 3);
-    }
+    /// Sets a vertex' generated tangent with the bi-tangent encoded as the W component in the
+    /// tangent. The W component marks if the bi-tangent is flipped. This will only be called if
+    /// `set_tangent` is not implemented.
+    fn set_tangent_encoded(&mut self, _tangent: Vector4<f32>, _face: usize, _vert: usize) {}
 }
 
-/// Returns the normal of the referenced face of vertex number
-/// iVert, in the range {0,1,2} for triangles, and {0,1,2,3} for quads.
-extern "C" fn normal(
-    pContext: *const ffi::SMikkTSpaceContext,
-    fvNormOut: *mut c_float,
-    iFace: c_int,
-    iVert: c_int,
-) {
-    unsafe {
-        let x = (*pContext).m_pUserData as *const Closures;
-        let slice = ((*x).normal)(iFace as usize, iVert as usize);
-        let src = slice.as_ptr() as *const c_float;
-        ptr::copy_nonoverlapping::<c_float>(src, fvNormOut, 3);
-    }
+/// Default (recommended) Angular Threshold is 180 degrees, which means threshold disabled.
+pub fn generate_tangents_default<I: Geometry>(geometry: &mut I) -> bool {
+    unsafe { generated::genTangSpace(geometry, 180.0) }
 }
 
-/// Returns the texcoord of the referenced face of vertex number
-/// iVert, in the range {0,1,2} for triangles, and {0,1,2,3} for quads.
-extern "C" fn tex_coord(
-    pContext: *const ffi::SMikkTSpaceContext,
-    fvTexcOut: *mut c_float,
-    iFace: c_int,
-    iVert: c_int,
-) {
-    unsafe {
-        let x = (*pContext).m_pUserData as *const Closures;
-        let slice = ((*x).tex_coord)(iFace as usize, iVert as usize);
-        let src = slice.as_ptr() as *const c_float;
-        ptr::copy_nonoverlapping::<c_float>(src, fvTexcOut, 2);
-    }
+fn get_position<I: Geometry>(geometry: &mut I, index: usize) -> Vector3<f32> {
```

Comment:
In the back of my mind I'm worried about code bloat with the generic parameter. It should be OK since the user is unlikely to implement `Geometry` for more than one data structure in their project but it is something to be aware of. I don't want to see a repeat of https://github.com/gltf-rs/gltf/issues/133! :smile:

Reply:
I chose to go with a generic parameter, because it allows the compiler to in-line the simple accessors.